### PR TITLE
Merge from moe_writing_branch_from_02a32292dab18eafe393676bf6d3a32725219078

### PR DESCRIPTION
**Pull request by cgruber**
Merge in change from google internal repository.  Notable large changes are:

  * Introduce Producers core API.  Producers are an asynchronous dependency graph suitable for creating worker graphs.  More detail on those will be added to documentation later.
  * Scope validations between components is stricter.
    - Singleton components cannot depend on non-Singleton components.
    - Custom scopes (always supported) are now required for multiple-scope situations
    - Inter-component scope relationships are evaluated to ensure they don't circle back (Singleton -> Request -> Singleton).
    - Some scope validation can be turned down to warning or off to permit Dagger1 project migration (see ComponentProcessor javadoc for options)
  * Some bug-fixes and cleanup

Note: Some regressions have been noted - notably with dependency cycle reporting.  If you get a stack overflow, you have a dependency cycle that is blowing up our reporting.  It's a known error we're fixing.

### CODE REVIEW

**Comment by adriancole** [comment]
:ramen: on me to whoever merges this!

File: producers/src/main/java/dagger/producers/Produces.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * a {@link ListenableFuture}, then the parameter type of the future is bound to the value that the
+ * future provides; otherwise, the return type is bound to the returned value. The production
+ * component will pass dependencies to the method as parameters.
+ *
+ * @author Jesse Beder
+ */
+@Documented @Target(METHOD)
+public @interface Produces {
+  /** The type of binding into which the return type of the annotated method contributes. */
+  enum Type {
```
Comment (gk5885) [inline producers/src/main/java/dagger/producers/Produces.java]: Right.  This has just been a migration headache that I haven't been itching to undertake.

**Comment by Amyntas** [comment]
Any hints on when this will be merged in ?

**Comment by cgruber** [comment]
I've re-done this sync, with all the additional changes - I was waiting on a review from @gk5885 originally and it got lost in the holiday shuffle.  This should be synced shortly. In the mean-time, I'm going to try for a weekly or more-frequently sync schedule.

**Comment by cgruber** [comment]
@gk5885 - ok - Travis is stable - can you PTAL?

**Comment by gk5885** [comment]
LGTM

**Pull request merged**